File: android/EGETrainerApp/core/src/main/kotlin/ru/egetrainer/core/services/SupabaseClient.kt
package ru.egetrainer.core.services

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import ru.egetrainer.core.AppJson
import ru.egetrainer.core.models.AuthSession
import java.io.IOException
import java.io.InterruptedIOException
import java.net.SocketTimeoutException
import java.util.concurrent.TimeUnit

/**
 * Единый сетевой слой Supabase — порт SupabaseClient.swift (зеркало
 * app/providers/supabase.js + supabase-rest.js):
 * - GoTrue: password-login, refresh, logout;
 * - PostgREST: rpc / select / insert;
 * - авто-refresh access_token (< 60 с до истечения) и один повтор запроса на 401;
 * - ретраи ТОЛЬКО сетевых сбоев/таймаутов (любой HTTP-ответ не ретраится),
 *   backoff как в вебе: 350 / 800 / 1500 мс.
 */
class SupabaseClient(private val store: SessionStore) {

    private val http: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .writeTimeout(20, TimeUnit.SECONDS)
        .callTimeout(20, TimeUnit.SECONDS)
        .build()

    private val sessionMutex = Mutex()

    @Volatile
    private var session: AuthSession? = store.load()

    private var refreshInFlight: CompletableDeferred<AuthSession?>? = null

    private val jsonMedia = "application/json".toMediaType()

    // MARK: Session

    val currentSession: AuthSession? get() = session

    suspend fun restoreSession(): AuthSession? {
        val s = session ?: return null
        if (!s.isExpiringSoon) return s
        return refreshSession()
    }

    fun setSession(s: AuthSession?) {
        session = s
        if (s != null) store.save(s) else store.clear()
    }

    // MARK: Auth (GoTrue)

    suspend fun signIn(email: String, password: String): AuthSession {
        val url = "${SupabaseConfig.BASE_URL}/auth/v1/token".toHttpUrl()
            .newBuilder().addQueryParameter("grant_type", "password").build()
        val body = JsonObject(
            mapOf("email" to JsonPrimitive(email), "password" to JsonPrimitive(password))
        ).toString()
        val req = Request.Builder().url(url)
            .post(body.toRequestBody(jsonMedia))
            .header("apikey", SupabaseConfig.ANON_KEY)
            .header("Content-Type", "application/json")
            .build()

        val (data, status) = dataWithRetry(req)
        if (status != 200) throw SupabaseError.fromAuthBody(status, data)
        val s = decode(AuthSession.serializer(), data).normalized()
        setSession(s)
        return s
    }

    /**
     * Best-effort refresh (как в вебе): при неудаче сессию не трогаем —
     * протухший токен поймает 401-ветка конкретного запроса.
     * Конкурентные вызовы дедуплицируются (один сетевой refresh).
     */
    suspend fun refreshSession(): AuthSession? {
        val waitFor: CompletableDeferred<AuthSession?>
        val isOwner: Boolean
        sessionMutex.withLock {
            val existing = refreshInFlight
            if (existing != null) {
                waitFor = existing
                isOwner = false
            } else {
                val current = session
                if (current == null || current.refreshToken.isEmpty()) return null
                val d = CompletableDeferred<AuthSession?>()
                refreshInFlight = d
                waitFor = d
                isOwner = true
            }
        }
        if (!isOwner) return waitFor.await()

        val refreshed: AuthSession? = try {
            doRefresh(session!!.refreshToken)
        } catch (_: Throwable) {
            null
        }
        if (refreshed != null) setSession(refreshed)
        sessionMutex.withLock { refreshInFlight = null }
        waitFor.complete(refreshed)
        return refreshed
    }

    private suspend fun doRefresh(refreshToken: String): AuthSession? {
        val url = "${SupabaseConfig.BASE_URL}/auth/v1/token".toHttpUrl()
            .newBuilder().addQueryParameter("grant_type", "refresh_token").build()
        val body = JsonObject(mapOf("refresh_token" to JsonPrimitive(refreshToken))).toString()
        val req = Request.Builder().url(url)
            .post(body.toRequestBody(jsonMedia))
            .header("apikey", SupabaseConfig.ANON_KEY)
            .header("Content-Type", "application/json")
            .build()
        val (data, status) = runCatching { execute(req) }.getOrNull() ?: return null
        if (status != 200) return null
        return runCatching { decode(AuthSession.serializer(), data).normalized() }.getOrNull()
    }

    suspend fun signOut() {
        val s = session
        if (s != null) {
            val req = Request.Builder()
                .url("${SupabaseConfig.BASE_URL}/auth/v1/logout")
                .post(ByteArray(0).toRequestBody(null))
                .header("apikey", SupabaseConfig.ANON_KEY)
                .header("Authorization", "Bearer ${s.accessToken}")
                .build()
            runCatching { execute(req) } // best-effort revoke, UX не блокируем
        }
        setSession(null)
    }

    // MARK: PostgREST

    /** RPC-вызов: POST /rest/v1/rpc/<name>. Декодирует ответ в T. */
    suspend fun <T> rpc(
        name: String,
        params: JsonObject = JsonObject(emptyMap()),
        deserializer: DeserializationStrategy<T>,
    ): T = decode(deserializer, rpcRaw(name, params))

    /** RPC без интереса к телу ответа. */
    suspend fun rpcVoid(name: String, params: JsonObject = JsonObject(emptyMap())) {
        rpcRaw(name, params)
    }

    /** RPC, возвращающий сырое JSON-дерево (для форм «массив ИЛИ объект»). */
    suspend fun rpcElement(name: String, params: JsonObject = JsonObject(emptyMap())): JsonElement {
        val data = rpcRaw(name, params)
        return runCatching { AppJson.parseToJsonElement(data.decodeToString()) }
            .getOrElse { throw SupabaseError.Decoding(it.message ?: "parse") }
    }

    /**
     * PostgREST для scalar-RPC может вернуть как объект, так и массив из одной
     * строки — хелпер разворачивает обе формы (зеркало rpcSingleRow iOS /
     * normalizeAttemptRowFromRpc в вебе).
     */
    suspend fun <T> rpcSingleRow(
        name: String,
        params: JsonObject = JsonObject(emptyMap()),
        deserializer: DeserializationStrategy<T>,
    ): T {
        val element = rpcElement(name, params)
        val row: JsonElement = when (element) {
            is JsonArray -> element.firstOrNull() ?: throw SupabaseError.EmptyResponse()
            else -> element
        }
        return runCatching { AppJson.decodeFromJsonElement(deserializer, row) }
            .getOrElse { throw SupabaseError.Decoding(it.message ?: "decode") }
    }

    /** REST select: GET /rest/v1/<table>?<query> */
    suspend fun <T> select(
        table: String,
        query: Map<String, String>,
        deserializer: DeserializationStrategy<T>,
    ): T {
        val urlBuilder = "${SupabaseConfig.BASE_URL}/rest/v1/$table".toHttpUrl().newBuilder()
        query.forEach { (k, v) -> urlBuilder.addQueryParameter(k, v) }
        val req = Request.Builder().url(urlBuilder.build()).get().build()
        return decode(deserializer, authedRequest(req))
    }

    /** REST insert: POST /rest/v1/<table> (Prefer: return=representation). */
    suspend fun <T> insert(
        table: String,
        values: JsonObject,
        deserializer: DeserializationStrategy<T>,
    ): T {
        val req = Request.Builder()
            .url("${SupabaseConfig.BASE_URL}/rest/v1/$table")
            .post(values.toString().toRequestBody(jsonMedia))
            .header("Content-Type", "application/json")
            .header("Prefer", "return=representation")
            .build()
        return decode(deserializer, authedRequest(req))
    }

    // MARK: Internals

    private suspend fun rpcRaw(name: String, params: JsonObject): ByteArray {
        val req = Request.Builder()
            .url("${SupabaseConfig.BASE_URL}/rest/v1/rpc/$name")
            .post(params.toString().toRequestBody(jsonMedia))
            .header("Content-Type", "application/json")
            .build()
        return authedRequest(req)
    }

    /**
     * Запрос с auth-заголовками, авто-refresh при близком к истечению токене
     * и одним повтором на 401 (зеркало 401-retry из supabase-rest.js).
     */
    internal suspend fun authedRequest(request: Request): ByteArray {
        var s = session ?: throw SupabaseError.AuthRequired()
        if (s.isExpiringSoon) {
            refreshSession()?.let { s = it }
        }

        suspend fun attempt(token: String): Pair<ByteArray, Int> {
            val req = request.newBuilder()
                .header("apikey", SupabaseConfig.ANON_KEY)
                .header("Authorization", "Bearer $token")
                .build()
            return dataWithRetry(req)
        }

        var (data, status) = attempt(s.accessToken)
        if (status == 401) {
            val refreshed = refreshSession()
            if (refreshed == null) {
                setSession(null)
                throw SupabaseError.AuthRequired()
            }
            val second = attempt(refreshed.accessToken)
            data = second.first
            status = second.second
        }

        if (status !in 200..299) throw mapHTTPError(status, data)
        return data
    }

    /** Сетевые ретраи: повторяем только IOException (сбой/таймаут), любой HTTP-ответ возвращаем сразу. */
    internal suspend fun dataWithRetry(req: Request, retries: Int = 2): Pair<ByteArray, Int> {
        val backoffsMs = longArrayOf(350, 800, 1500)
        for (attemptIdx in 0..retries) {
            try {
                return execute(req)
            } catch (e: IOException) {
                val isTimeout = e is SocketTimeoutException || e is InterruptedIOException
                if (attemptIdx == retries) {
                    if (isTimeout) throw SupabaseError.Timeout()
                    throw SupabaseError.Network(e)
                }
                delay(backoffsMs[minOf(attemptIdx, backoffsMs.size - 1)])
            }
        }
        throw SupabaseError.Network(null)
    }

    private suspend fun execute(req: Request): Pair<ByteArray, Int> =
        withContext(Dispatchers.IO) {
            http.newCall(req).execute().use { resp ->
                Pair(resp.body?.bytes() ?: ByteArray(0), resp.code)
            }
        }

    private fun mapHTTPError(status: Int, body: ByteArray): SupabaseError {
        val obj = runCatching {
            AppJson.parseToJsonElement(body.decodeToString()) as? JsonObject
        }.getOrNull()
        val msg = (obj?.get("message") as? JsonPrimitive)
            ?.takeIf { it.isString }?.content ?: ""
        if (status == 403 || msg.uppercase().contains("ACCESS_DENIED")) return SupabaseError.AccessDenied()
        if (status == 401) return SupabaseError.AuthRequired()
        return SupabaseError.Http(status, msg)
    }

    internal fun <T> decode(deserializer: DeserializationStrategy<T>, data: ByteArray): T =
        try {
            AppJson.decodeFromString(deserializer, data.decodeToString())
        } catch (e: Exception) {
            throw SupabaseError.Decoding(e.toString())
        }

    internal fun baseHttpUrl(): HttpUrl = SupabaseConfig.BASE_URL.toHttpUrl()
}

File: ios/EGETrainerApp/EGETrainerApp/Services/SupabaseClient.swift
import Foundation

/// Единый сетевой слой Supabase (зеркало app/providers/supabase.js + supabase-rest.js):
/// - GoTrue: password-login, refresh, logout;
/// - PostgREST: rpc / select / insert;
/// - авто-refresh access_token и один повтор запроса на 401;
/// - ретраи ТОЛЬКО сетевых сбоев/таймаутов (любой HTTP-ответ не ретраится),
///   backoff как в вебе: 350 / 800 / 1500 мс.
actor SupabaseClient {
    static let shared = SupabaseClient(store: KeychainSessionStore())

    private let store: SessionStore
    private var session: AuthSession?
    private var refreshTask: Task<AuthSession?, Never>?

    private let urlSession: URLSession

    init(store: SessionStore) {
        self.store = store
        self.session = store.load()
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 20
        cfg.requestCachePolicy = .reloadIgnoringLocalCacheData
        self.urlSession = URLSession(configuration: cfg)
    }

    // MARK: - Session

    var currentSession: AuthSession? { session }

    func restoreSession() async -> AuthSession? {
        guard let s = session else { return nil }
        if !s.isExpiringSoon { return s }
        return await refreshSession()
    }

    func setSession(_ s: AuthSession?) {
        session = s
        if let s { store.save(s) } else { store.clear() }
    }

    // MARK: - Auth (GoTrue)

    func signIn(email: String, password: String) async throws -> AuthSession {
        var req = URLRequest(url: SupabaseConfig.baseURL.appendingPathComponent("auth/v1/token"))
        req.url = req.url?.appending(queryItems: [URLQueryItem(name: "grant_type", value: "password")])
        req.httpMethod = "POST"
        req.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["email": email, "password": password])

        let (data, resp) = try await dataWithRetry(req)
        guard let http = resp as? HTTPURLResponse else { throw SupabaseError.emptyResponse }
        guard http.statusCode == 200 else {
            throw SupabaseError.fromAuthBody(status: http.statusCode, body: data)
        }
        let s = try decode(AuthSession.self, from: data)
        setSession(s)
        return s
    }

    @discardableResult
    func refreshSession() async -> AuthSession? {
        if let task = refreshTask { return await task.value }
        guard let current = session, !current.refreshToken.isEmpty else { return nil }

        let task = Task<AuthSession?, Never> { [urlSession] in
            var req = URLRequest(url: SupabaseConfig.baseURL.appendingPathComponent("auth/v1/token"))
            req.url = req.url?.appending(queryItems: [URLQueryItem(name: "grant_type", value: "refresh_token")])
            req.httpMethod = "POST"
            req.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try? JSONEncoder().encode(["refresh_token": current.refreshToken])
            guard let (data, resp) = try? await urlSession.data(for: req),
                  let http = resp as? HTTPURLResponse, http.statusCode == 200,
                  let s = try? JSONDecoder().decode(AuthSession.self, from: data)
            else { return nil }
            return s
        }
        refreshTask = task
        let refreshed = await task.value
        refreshTask = nil
        // Refresh не удался — сессию не трогаем (best-effort, как в вебе);
        // протухший токен поймает 401-ветка конкретного запроса.
        if let refreshed { setSession(refreshed) }
        return refreshed
    }

    func signOut() async {
        if let s = session {
            var req = URLRequest(url: SupabaseConfig.baseURL.appendingPathComponent("auth/v1/logout"))
            req.httpMethod = "POST"
            req.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
            req.setValue("Bearer \(s.accessToken)", forHTTPHeaderField: "Authorization")
            _ = try? await urlSession.data(for: req) // best-effort revoke, UX не блокируем
        }
        setSession(nil)
    }

    // MARK: - PostgREST

    /// RPC-вызов: POST /rest/v1/rpc/<name>. Декодирует ответ в T.
    func rpc<T: Decodable>(_ name: String, params: [String: JSONValue] = [:], as type: T.Type) async throws -> T {
        let url = SupabaseConfig.baseURL.appendingPathComponent("rest/v1/rpc/\(name)")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(params)
        let data = try await authedRequest(req)
        return try decode(T.self, from: data)
    }

    /// RPC без интереса к телу ответа.
    func rpcVoid(_ name: String, params: [String: JSONValue] = [:]) async throws {
        let url = SupabaseConfig.baseURL.appendingPathComponent("rest/v1/rpc/\(name)")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(params)
        _ = try await authedRequest(req)
    }

    /// REST select: GET /rest/v1/<table>?<query>
    func select<T: Decodable>(_ table: String, query: [URLQueryItem], as type: T.Type) async throws -> T {
        var comps = URLComponents(url: SupabaseConfig.baseURL.appendingPathComponent("rest/v1/\(table)"),
                                  resolvingAgainstBaseURL: false)!
        comps.queryItems = query
        var req = URLRequest(url: comps.url!)
        req.httpMethod = "GET"
        let data = try await authedRequest(req)
        return try decode(T.self, from: data)
    }

    /// REST insert: POST /rest/v1/<table> (Prefer: return=representation).
    func insert<T: Decodable>(_ table: String, values: [String: JSONValue], as type: T.Type) async throws -> T {
        var req = URLRequest(url: SupabaseConfig.baseURL.appendingPathComponent("rest/v1/\(table)"))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("return=representation", forHTTPHeaderField: "Prefer")
        req.httpBody = try JSONEncoder().encode(values)
        let data = try await authedRequest(req)
        return try decode(T.self, from: data)
    }

    // MARK: - Internals

    /// Запрос с auth-заголовками, авто-refresh при близком к истечению токене
    /// и одним повтором на 401 (зеркало 401-retry из supabase-rest.js).
    private func authedRequest(_ request: URLRequest) async throws -> Data {
        guard var s = session else { throw SupabaseError.authRequired }
        if s.isExpiringSoon {
            if let refreshed = await refreshSession() { s = refreshed }
        }

        func attempt(token: String) async throws -> (Data, Int) {
            var req = request
            req.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            let (data, resp) = try await dataWithRetry(req)
            let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
            return (data, status)
        }

        var (data, status) = try await attempt(token: s.accessToken)
        if status == 401 {
            guard let refreshed = await refreshSession() else {
                setSession(nil)
                throw SupabaseError.authRequired
            }
            (data, status) = try await attempt(token: refreshed.accessToken)
        }

        guard (200..<300).contains(status) else {
            throw mapHTTPError(status: status, body: data)
        }
        return data
    }

    /// Сетевые ретраи: повторяем только URLError (сбой/таймаут), любой HTTP-ответ возвращаем сразу.
    func dataWithRetry(_ req: URLRequest, retries: Int = 2) async throws -> (Data, URLResponse) {
        let backoffsMs: [UInt64] = [350, 800, 1500]
        var lastError: Error = SupabaseError.timeout
        for attemptIdx in 0...retries {
            do {
                return try await urlSession.data(for: req)
            } catch {
                lastError = error
                let isRetryable = (error as? URLError).map {
                    [.timedOut, .cannotConnectToHost, .networkConnectionLost,
                     .notConnectedToInternet, .dnsLookupFailed, .cannotFindHost].contains($0.code)
                } ?? false
                if !isRetryable || attemptIdx == retries {
                    if let urlErr = error as? URLError, urlErr.code == .timedOut {
                        throw SupabaseError.timeout
                    }
                    throw SupabaseError.network(underlying: error)
                }
                let delay = backoffsMs[min(attemptIdx, backoffsMs.count - 1)]
                try? await Task.sleep(nanoseconds: delay * 1_000_000)
            }
        }
        throw SupabaseError.network(underlying: lastError)
    }

    private func mapHTTPError(status: Int, body: Data) -> SupabaseError {
        struct PgErr: Decodable { let message: String?; let hint: String?; let details: String? }
        let parsed = try? JSONDecoder().decode(PgErr.self, from: body)
        let msg = parsed?.message ?? ""
        if status == 403 || msg.uppercased().contains("ACCESS_DENIED") { return .accessDenied }
        if status == 401 { return .authRequired }
        return .http(status: status, message: msg)
    }

    func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw SupabaseError.decoding(message: String(describing: error))
        }
    }
}

extension SupabaseClient {
    /// PostgREST для scalar-RPC может вернуть как объект, так и массив из одной строки —
    /// хелпер разворачивает обе формы (зеркало normalizeAttemptRowFromRpc в вебе).
    func rpcSingleRow<T: Decodable>(_ name: String, params: [String: JSONValue] = [:], as type: T.Type) async throws -> T {
        if let rows = try? await rpc(name, params: params, as: [T].self) {
            if let first = rows.first { return first }
            throw SupabaseError.emptyResponse
        }
        return try await rpc(name, params: params, as: T.self)
    }
}
